radicle-tracker/src/thread.rs

```diff
@@ -0,0 +1,380 @@
+use nonempty::NonEmpty;
+use rose_tree::{NodeIndex, RoseTree, ROOT};
+use std::{collections::HashMap, hash};
+
+/// Laws:
+/// Thread::new(comment).first() === comment
+/// thread.reply(comment).delete(comment) === thread
+/// Thread::new(comment).delete(comment) === None
+/// Thread::new(comment).edit(f, comment) === Thread::new(f(comment).unwrap())
+#[derive(Debug, Clone)]
+pub struct Thread<A> {
+    lut: HashMap<Path, NodeIndex>,
+    tree: RoseTree<A>,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum Error {
+    MissingPath(Path),
+    MissingNode,
+}
+
+#[derive(Debug, PartialEq, Eq, hash::Hash, Clone)]
+pub struct Path(Vec<u32>);
```

Comment:
Could you elaborate a bit on your choice of datastructure, and what might be possible alternatives?

The `Path` type in particular strikes me as a little odd. My mental model of a "thread" (which, admittedly, originates in the 90s) _basically_ only considers the parent -- I'm looking at an entry, and then may decide to reply to it, or I'm interested in the replies to it. If you consider a pathologically long thread, I might want to _collapse_ nodes below a certain threshold, or "pin" a certain node at the root.

In other words, is navigating a tree structure not more like a zipper?

Reply:
Mmmm that could be a good thing to explore.

So you're saying that we could have:
```haskell
Thread = (Thread, [Comment], Thread)

reply :: Comment -> Thread -> Thread
reply comment (left, comments, right) = (left, comments <> [comment], right)
```

But hmmmm... I still think you need a way to "navigate" to a thread. I can think over this a bit and give more elaboration to my choice. We might want to pair on the possible alternatives, since I'm not entirely sure what there is :smile: The `Zipper` version looks isomorphic to the RoseTree version, and at the moment not entirely sure what it buys us; but at the same time I'd need to explore the rest of the functionality more.

Reply:
>  I still think you need a way to "navigate" to a thread

Sure... but isn't it like browsing any tree-shaped structure? Like, you know a root, list the outgoing children, then focus on the one you're interested in, etc. An absolute path is potentially useful as a "permalink" sort of thing -- but then, you can also address it as `root,left,left,right,left,left` etc.

A zipper representation seems to also map much more obviously to a partially-materialised view (i.e. you don't have to keep the entire tree in memory).

Reply:
Ya great point, lemme work with it!